File: src/connectors/slack.ts
import { readFileSync, existsSync, readdirSync } from "fs";
import { join, extname } from "path";
import { execSync } from "child_process";
import type { RawMessage } from "../corpus/types.js";

// Slack workspace export format:
// ZIP containing directories per channel, each with JSON files.
// Each JSON file is an array of message objects:
// { type: "message", user: "U...", text: "...", ts: "1234567890.123456" }

interface SlackMessage {
  type?: string;
  subtype?: string;
  user?: string;
  text?: string;
  ts?: string;
  thread_ts?: string;
}

export interface SlackImportOptions {
  file: string; // path to ZIP or extracted directory
  userId?: string; // Slack user ID to filter (if known)
  userName?: string; // display name to filter from users.json
  since?: Date;
}

export function importSlack(opts: SlackImportOptions): RawMessage[] {
  let dir = opts.file;

  // if it's a zip, extract to tmp
  if (opts.file.endsWith(".zip")) {
    dir = "/tmp/replicateme-slack-export";
    execSync(`rm -rf ${dir} && mkdir -p ${dir} && unzip -q -o "${opts.file}" -d ${dir}`);
  }

  if (!existsSync(dir)) {
    throw new Error(`Slack export not found at ${dir}`);
  }

  // try to resolve user ID from users.json
  let userId = opts.userId;
  if (!userId && opts.userName) {
    const usersFile = join(dir, "users.json");
    if (existsSync(usersFile)) {
      const users = JSON.parse(readFileSync(usersFile, "utf-8")) as Array<{
        id: string;
        name: string;
        real_name?: string;
        profile?: { display_name?: string };
      }>;
      const match = users.find(
        (u) =>
          u.name === opts.userName ||
          u.real_name === opts.userName ||
          u.profile?.display_name === opts.userName
      );
      if (match) userId = match.id;
    }
  }

  const messages: RawMessage[] = [];
  const channels = readdirSync(dir, { withFileTypes: true })
    .filter((d) => d.isDirectory())
    .map((d) => d.name);

  for (const channel of channels) {
    const channelDir = join(dir, channel);
    const jsonFiles = readdirSync(channelDir).filter(
      (f) => extname(f) === ".json"
    );

    for (const file of jsonFiles) {
      const raw = readFileSync(join(channelDir, file), "utf-8");
      let msgs: SlackMessage[];
      try {
        msgs = JSON.parse(raw);
      } catch {
        continue;
      }

      for (const msg of msgs) {
        if (msg.type !== "message" || msg.subtype) continue;
        if (!msg.text || !msg.ts) continue;
        if (userId && msg.user !== userId) continue;

        const timestamp = new Date(parseFloat(msg.ts) * 1000);
        if (opts.since && timestamp < opts.since) continue;

        messages.push({
          id: `slack-${msg.ts}`,
          text: cleanSlackText(msg.text),
          platform: "slack",
          timestamp,
          isFromUser: true,
          metadata: {
            channel,
            threadTs: msg.thread_ts,
            userId: msg.user,
          },
        });
      }
    }
  }

  messages.sort((a, b) => a.timestamp.getTime() - b.timestamp.getTime());
  return messages;
}

function cleanSlackText(text: string): string {
  // strip user mentions <@U123> -> @user
  let cleaned = text.replace(/<@[A-Z0-9]+>/g, "@user");
  // strip channel refs <#C123|channel> -> #channel
  cleaned = cleaned.replace(/<#[A-Z0-9]+\|([^>]+)>/g, "#$1");
  // strip URLs <http://...|label> -> label or url
  cleaned = cleaned.replace(/<(https?:\/\/[^|>]+)\|([^>]+)>/g, "$2");
  cleaned = cleaned.replace(/<(https?:\/\/[^>]+)>/g, "$1");
  return cleaned;
}

File: pkg/connectors/slack.go
package connectors

import (
	"archive/zip"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/jschell12/replicateme/pkg/corpus"
)

// SlackImportOptions controls Slack workspace export import.
type SlackImportOptions struct {
	File     string // path to ZIP or extracted directory
	UserID   string // Slack user ID to filter
	UserName string // display name to resolve via users.json
	Since    *time.Time
}

type slackMessage struct {
	Type     string `json:"type"`
	Subtype  string `json:"subtype,omitempty"`
	User     string `json:"user,omitempty"`
	Text     string `json:"text,omitempty"`
	TS       string `json:"ts,omitempty"`
	ThreadTS string `json:"thread_ts,omitempty"`
}

type slackUser struct {
	ID       string `json:"id"`
	Name     string `json:"name"`
	RealName string `json:"real_name,omitempty"`
	Profile  struct {
		DisplayName string `json:"display_name,omitempty"`
	} `json:"profile,omitempty"`
}

// ImportSlack reads messages from a Slack workspace export.
func ImportSlack(opts SlackImportOptions) ([]corpus.RawMessage, error) {
	dir := opts.File

	if strings.HasSuffix(opts.File, ".zip") {
		dir = "/tmp/replicateme-slack-export"
		if err := extractZip(opts.File, dir); err != nil {
			return nil, fmt.Errorf("extract slack zip: %w", err)
		}
	}

	if _, err := os.Stat(dir); err != nil {
		return nil, fmt.Errorf("Slack export not found at %s", dir)
	}

	// resolve user ID from users.json if needed
	userID := opts.UserID
	if userID == "" && opts.UserName != "" {
		usersFile := filepath.Join(dir, "users.json")
		if data, err := os.ReadFile(usersFile); err == nil {
			var users []slackUser
			if json.Unmarshal(data, &users) == nil {
				for _, u := range users {
					if u.Name == opts.UserName || u.RealName == opts.UserName || u.Profile.DisplayName == opts.UserName {
						userID = u.ID
						break
					}
				}
			}
		}
	}

	var messages []corpus.RawMessage

	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil, err
	}

	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		channel := entry.Name()
		channelDir := filepath.Join(dir, channel)

		files, err := os.ReadDir(channelDir)
		if err != nil {
			continue
		}
		for _, f := range files {
			if filepath.Ext(f.Name()) != ".json" {
				continue
			}
			data, err := os.ReadFile(filepath.Join(channelDir, f.Name()))
			if err != nil {
				continue
			}
			var msgs []slackMessage
			if json.Unmarshal(data, &msgs) != nil {
				continue
			}

			for _, msg := range msgs {
				if msg.Type != "message" || msg.Subtype != "" {
					continue
				}
				if msg.Text == "" || msg.TS == "" {
					continue
				}
				if userID != "" && msg.User != userID {
					continue
				}

				var tsFloat float64
				fmt.Sscanf(msg.TS, "%f", &tsFloat)
				ts := time.Unix(int64(tsFloat), int64((tsFloat-float64(int64(tsFloat)))*1e9)).UTC()

				if opts.Since != nil && ts.Before(*opts.Since) {
					continue
				}

				messages = append(messages, corpus.RawMessage{
					ID:         fmt.Sprintf("slack-%s", msg.TS),
					Text:       cleanSlackText(msg.Text),
					Platform:   corpus.Slack,
					Timestamp:  ts,
					IsFromUser: true,
					Metadata: map[string]any{
						"channel":  channel,
						"threadTs": msg.ThreadTS,
						"userId":   msg.User,
					},
				})
			}
		}
	}

	sort.Slice(messages, func(i, j int) bool {
		return messages[i].Timestamp.Before(messages[j].Timestamp)
	})
	return messages, nil
}

var (
	slackUserMentionRe   = regexp.MustCompile(`<@[A-Z0-9]+>`)
	slackChannelRefRe    = regexp.MustCompile(`<#[A-Z0-9]+\|([^>]+)>`)
	slackURLWithLabelRe  = regexp.MustCompile(`<(https?://[^|>]+)\|([^>]+)>`)
	slackURLBareRe       = regexp.MustCompile(`<(https?://[^>]+)>`)
)

func cleanSlackText(text string) string {
	cleaned := slackUserMentionRe.ReplaceAllString(text, "@user")
	cleaned = slackChannelRefRe.ReplaceAllString(cleaned, "#$1")
	cleaned = slackURLWithLabelRe.ReplaceAllString(cleaned, "$2")
	cleaned = slackURLBareRe.ReplaceAllString(cleaned, "$1")
	return cleaned
}

// extractZip extracts a ZIP archive to dst, clearing dst first.
func extractZip(src, dst string) error {
	os.RemoveAll(dst)
	if err := os.MkdirAll(dst, 0o755); err != nil {
		return err
	}

	r, err := zip.OpenReader(src)
	if err != nil {
		return err
	}
	defer r.Close()

	for _, f := range r.File {
		path := filepath.Join(dst, f.Name)
		if !strings.HasPrefix(path, filepath.Clean(dst)+string(os.PathSeparator)) {
			continue // zip slip protection
		}
		if f.FileInfo().IsDir() {
			os.MkdirAll(path, f.Mode())
			continue
		}
		if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
			return err
		}
		rc, err := f.Open()
		if err != nil {
			return err
		}
		out, err := os.Create(path)
		if err != nil {
			rc.Close()
			return err
		}
		_, err = io.Copy(out, rc)
		rc.Close()
		out.Close()
		if err != nil {
			return err
		}
	}
	return nil
}
